File: backend/domain_monitor/domain_sources.py
from __future__ import annotations

import logging
from dataclasses import dataclass, field
from pathlib import Path
from typing import Iterable, Iterator, Optional

import crawl4ai_source
import source_adapters
import source_config
from models import normalize_domain
from source_adapters import (
    STATUS_ACTIVE,
    STATUS_CONFIGURED,
    STATUS_DISABLED,
    STATUS_FAILED,
    STATUS_NOT_CONFIGURED,
    DomainSource,
    InlineListSource,
    SourceReport,
)
from source_config import SourceSettings, load_settings

logger = logging.getLogger(__name__)

# Re-exported for callers that only need the manual folder.
SOURCES_DIR = source_config.MODULE_DIR / "sources"


@dataclass
class CollectionResult:
    """Outcome of one discovery pass.

    `discovered` counts every raw string an adapter produced, `valid` counts
    those that survived normalisation and `unique` is what actually reaches
    RDAP. The three are reported separately because the gaps between them are
    the interesting part.
    """

    domains: list[str] = field(default_factory=list)
    origins: dict[str, str] = field(default_factory=dict)
    reports: list[SourceReport] = field(default_factory=list)
    discovered: int = 0
    valid: int = 0
    invalid: int = 0
    duplicates: int = 0
    truncated: bool = False

    @property
    def unique(self) -> int:
        return len(self.domains)

    @property
    def any_source_configured(self) -> bool:
        return any(report.configured for report in self.reports)

    def per_source(self) -> dict[str, int]:
        return {report.name: report.raw_count for report in self.reports}


def build_sources(
    settings: Optional[SourceSettings] = None,
    kinds: Optional[Iterable[str]] = None,
) -> tuple[list[DomainSource], SourceSettings]:
    """Instantiate the enabled adapters.

    `kinds` lets a scan request narrow the selection; anything not enabled by
    configuration stays out regardless of what the request asks for.
    """
    settings = settings or load_settings()
    requested = {k.lower() for k in kinds} if kinds else None

    sources: list[DomainSource] = []
    for kind in source_config.ALL_KINDS:
        if not settings.is_enabled(kind):
            continue
        if requested is not None and kind not in requested:
            continue
        sources.extend(source_adapters.build_adapters(kind, settings))
    return sources, settings


def source_status(settings: Optional[SourceSettings] = None) -> list[dict]:
    """Per-source status for the Data Sources panel.

    Reports configuration state only; it never fetches, so it is cheap enough
    to call on every dashboard load.
    """
    settings = settings or load_settings()
    rows = []
    for adapter in source_adapters.all_adapters(settings):
        enabled = settings.is_enabled(adapter.kind)
        configured = adapter.is_configured()
        if not enabled:
            status = STATUS_DISABLED
        elif not configured:
            status = STATUS_NOT_CONFIGURED
        else:
            status = STATUS_CONFIGURED
        rows.append(
            {
                "kind": adapter.kind,
                "name": adapter.name,
                "label": adapter.label,
                "status": status,
                "enabled": enabled,
                "configured": configured,
                "detail": adapter.describe(),
            }
        )
    if settings.is_enabled(source_config.KIND_CRAWL4AI) or crawl4ai_source.load_source_configs():
        rows.extend(crawl4ai_source.source_status_rows())
    return rows


def collect(
    sources: Iterable[DomainSource],
    settings: Optional[SourceSettings] = None,
) -> CollectionResult:
    """Pull every source, normalise, deduplicate and cap.

    Deduplication happens here, before any RDAP call, so www.example.com and
    https://EXAMPLE.com/ cost exactly one lookup between them. A source that
    raises is marked Failed and the remaining sources still run.
    """
    settings = settings or load_settings()
    result = CollectionResult()
    seen: set[str] = set()
    cap = settings.max_candidates

    for source in sources:
        enabled = settings.is_enabled(getattr(source, "kind", "inline")) or isinstance(
            source, InlineListSource
        )
        configured = source.is_configured()
        report = SourceReport(
            kind=getattr(source, "kind", "inline"),
            name=source.name,
            label=source.label,
            status=STATUS_ACTIVE,
            configured=configured,
            enabled=enabled,
            detail=source.describe(),
        )

        if not configured:
            report.status = STATUS_NOT_CONFIGURED
            result.reports.append(report)
            logger.info("[source:%s] not configured — %s", source.name, report.detail)
            continue

        raw = 0
        accepted = 0
        try:
            for candidate in source.fetch_domains():
                raw += 1
                result.discovered += 1

                domain = normalize_domain(candidate)
                if not domain:
                    result.invalid += 1
                    continue
                result.valid += 1

                if domain in seen:
                    result.duplicates += 1
                    continue

                if len(result.domains) >= cap:
                    result.truncated = True
                    break

                seen.add(domain)
                result.domains.append(domain)
                result.origins[domain] = source.name
                accepted += 1
        except Exception as exc:
            # A broken source degrades to Failed; discovery continues.
            report.status = STATUS_FAILED
            report.error = str(exc)
            logger.warning("[source:%s] FAILED: %s", source.name, exc)
        else:
            logger.info(
                "[source:%s] %d raw, %d new candidates%s",
                source.name,
                raw,
                accepted,
                " (candidate cap reached)" if result.truncated else "",
            )

        report.raw_count = raw
        result.reports.append(report)

        if result.truncated:
            logger.warning(
                "[collect] candidate cap of %d reached; remaining sources skipped", cap
            )
            break

    logger.info(
        "[normalize] %d valid of %d discovered (%d rejected)",
        result.valid,
        result.discovered,
        result.invalid,
    )
    logger.info(
        "[dedupe] %d unique candidates (%d duplicates collapsed)",
        result.unique,
        result.duplicates,
    )
    return result


def parse_candidate_lines(text: str, column: int = 0) -> Iterator[str]:
    """Yield raw candidate cells from a pasted or uploaded TXT/CSV blob.

    Shared by the import endpoint and the file-backed adapters so both accept
    exactly the same formats.
    """
    for line in (text or "").splitlines():
        yield from source_adapters._cells(line, column)


def manual_import_path(settings: Optional[SourceSettings] = None) -> Path:
    """Where dashboard imports are appended, so ManualFileSource can re-read them."""
    settings = settings or load_settings()
    root = settings.manual_dir
    root.mkdir(parents=True, exist_ok=True)
    return root / Path(settings.manual_file).name


# --- Backwards-compatible helpers ------------------------------------------

def load_configured_sources() -> list[DomainSource]:
    sources, _ = build_sources()
    return sources


def collect_candidates(
    sources: Iterable[DomainSource],
) -> tuple[list[str], dict[str, int]]:
    """Legacy shape: (domains, per-source counts). Prefer `collect`."""
    result = collect(sources)
    collect_candidates.last_origin = result.origins  # type: ignore[attr-defined]
    return result.domains, result.per_source()

File: node-backend/domainMonitor/domainSources.js
"use strict";

// Faithful Node port of backend/domain_monitor/domain_sources.py
const fs = require("fs");
const path = require("path");

const sourceConfig = require("./sourceConfig");
const sourceAdapters = require("./sourceAdapters");
const { normalizeDomain } = require("./models");
const {
  STATUS_ACTIVE,
  STATUS_CONFIGURED,
  STATUS_DISABLED,
  STATUS_FAILED,
  STATUS_NOT_CONFIGURED,
  InlineListSource,
  SourceReport,
} = sourceAdapters;
const { loadSettings } = sourceConfig;

// crawl4aiSource is required lazily so this module loads even if that port
// has not been written yet.
function crawl4ai() {
  return require("./crawl4aiSource");
}

// Re-exported for callers that only need the manual folder.
const SOURCES_DIR = path.join(sourceConfig.MODULE_DIR, "sources");

/**
 * Outcome of one discovery pass.
 *
 * `discovered` counts every raw string an adapter produced, `valid` counts
 * those that survived normalisation and `unique` is what actually reaches
 * RDAP. The three are reported separately because the gaps between them are
 * the interesting part.
 */
class CollectionResult {
  constructor() {
    this.domains = [];
    this.origins = {};
    this.reports = [];
    this.discovered = 0;
    this.valid = 0;
    this.invalid = 0;
    this.duplicates = 0;
    this.truncated = false;
  }

  get unique() {
    return this.domains.length;
  }

  get any_source_configured() {
    return this.reports.some((report) => report.configured);
  }

  perSource() {
    const out = {};
    for (const report of this.reports) {
      out[report.name] = report.raw_count;
    }
    return out;
  }
}

/**
 * Instantiate the enabled adapters.
 *
 * `kinds` lets a scan request narrow the selection; anything not enabled by
 * configuration stays out regardless of what the request asks for.
 */
function buildSources(settings = null, kinds = null) {
  settings = settings || loadSettings();
  const requested = kinds
    ? new Set(Array.from(kinds).map((k) => k.toLowerCase()))
    : null;

  const sources = [];
  for (const kind of sourceConfig.ALL_KINDS) {
    if (!settings.isEnabled(kind)) {
      continue;
    }
    if (requested !== null && !requested.has(kind)) {
      continue;
    }
    sources.push(...sourceAdapters.buildAdapters(kind, settings));
  }
  return [sources, settings];
}

/**
 * Per-source status for the Data Sources panel.
 *
 * Reports configuration state only; it never fetches, so it is cheap enough
 * to call on every dashboard load.
 */
function sourceStatus(settings = null) {
  settings = settings || loadSettings();
  const rows = [];
  for (const adapter of sourceAdapters.allAdapters(settings)) {
    const enabled = settings.isEnabled(adapter.kind);
    const configured = adapter.isConfigured();
    let status;
    if (!enabled) {
      status = STATUS_DISABLED;
    } else if (!configured) {
      status = STATUS_NOT_CONFIGURED;
    } else {
      status = STATUS_CONFIGURED;
    }
    rows.push({
      kind: adapter.kind,
      name: adapter.name,
      label: adapter.label,
      status,
      enabled,
      configured,
      detail: adapter.describe(),
    });
  }
  if (settings.isEnabled(sourceConfig.KIND_CRAWL4AI) || crawl4ai().loadSourceConfigs().length) {
    rows.push(...crawl4ai().sourceStatusRows());
  }
  return rows;
}

/**
 * Pull every source, normalise, deduplicate and cap.
 *
 * Deduplication happens here, before any RDAP call, so www.example.com and
 * https://EXAMPLE.com/ cost exactly one lookup between them. A source that
 * raises is marked Failed and the remaining sources still run.
 */
async function collect(sources, settings = null) {
  settings = settings || loadSettings();
  const result = new CollectionResult();
  const seen = new Set();
  const cap = settings.max_candidates;

  for (const source of sources) {
    const kind = source.kind !== undefined ? source.kind : "inline";
    const enabled =
      settings.isEnabled(kind) || source instanceof InlineListSource;
    const configured = source.isConfigured();
    const report = new SourceReport({
      kind,
      name: source.name,
      label: source.label,
      status: STATUS_ACTIVE,
      configured,
      enabled,
      detail: source.describe(),
    });

    if (!configured) {
      report.status = STATUS_NOT_CONFIGURED;
      result.reports.push(report);
      console.info(`[source:${source.name}] not configured — ${report.detail}`);
      continue;
    }

    let raw = 0;
    let accepted = 0;
    let failed = false;
    try {
      for await (const candidate of source.fetchDomains()) {
        raw += 1;
        result.discovered += 1;

        const domain = normalizeDomain(candidate);
        if (!domain) {
          result.invalid += 1;
          continue;
        }
        result.valid += 1;

        if (seen.has(domain)) {
          result.duplicates += 1;
          continue;
        }

        if (result.domains.length >= cap) {
          result.truncated = true;
          break;
        }

        seen.add(domain);
        result.domains.push(domain);
        result.origins[domain] = source.name;
        accepted += 1;
      }
    } catch (exc) {
      // A broken source degrades to Failed; discovery continues.
      failed = true;
      report.status = STATUS_FAILED;
      report.error = String(exc && exc.message ? exc.message : exc);
      console.warn(`[source:${source.name}] FAILED: ${exc}`);
    }
    if (!failed) {
      console.info(
        `[source:${source.name}] ${raw} raw, ${accepted} new candidates${
          result.truncated ? " (candidate cap reached)" : ""
        }`
      );
    }

    report.raw_count = raw;
    result.reports.push(report);

    if (result.truncated) {
      console.warn(
        `[collect] candidate cap of ${cap} reached; remaining sources skipped`
      );
      break;
    }
  }

  console.info(
    `[normalize] ${result.valid} valid of ${result.discovered} discovered (${result.invalid} rejected)`
  );
  console.info(
    `[dedupe] ${result.unique} unique candidates (${result.duplicates} duplicates collapsed)`
  );
  return result;
}

/**
 * Yield raw candidate cells from a pasted or uploaded TXT/CSV blob.
 *
 * Shared by the import endpoint and the file-backed adapters so both accept
 * exactly the same formats.
 */
function* parseCandidateLines(text, column = 0) {
  for (const line of (text || "").split(/\r?\n/)) {
    yield* sourceAdapters._cells(line, column);
  }
}

/** Where dashboard imports are appended, so ManualFileSource can re-read them. */
function manualImportPath(settings = null) {
  settings = settings || loadSettings();
  const root = settings.manual_dir;
  fs.mkdirSync(root, { recursive: true });
  return path.join(root, path.basename(settings.manual_file));
}

// --- Backwards-compatible helpers ------------------------------------------

function loadConfiguredSources() {
  const [sources] = buildSources();
  return sources;
}

/** Legacy shape: [domains, per-source counts]. Prefer `collect`. */
async function collectCandidates(sources) {
  const result = await collect(sources);
  collectCandidates.last_origin = result.origins;
  return [result.domains, result.perSource()];
}

module.exports = {
  SOURCES_DIR,
  CollectionResult,
  buildSources,
  sourceStatus,
  collect,
  parseCandidateLines,
  manualImportPath,
  loadConfiguredSources,
  collectCandidates,
};
